### PR TITLE
Fix race condition when connecting coincident holes

### LINKED ISSUE #119: Another race condition with coincident holes
#112 introduced a new failing case which was working before:

```js
[
[[4116,-20],[4116,-8],[4096,-8],[4096,223],[3950,223],[3950,685],[3659,685],[3659,800],[3368,800],[3368,916],[3222,916],[3222,1031],[2931,1031],[2931,1147],[2785,1147],[2785,1031],[2640,1031],[2640,1262],[2931,1262],[2931,1378],[2494,1378],[2494,1493],[2348,1493],[2348,1609],[2203,1609],[2203,1839],[2057,1839],[2057,1955],[1911,1955],[1911,2185],[1766,2185],[1766,2532],[1620,2532],[1620,2647],[1475,2647],[1475,2877],[1183,2877],[1183,2993],[1038,2993],[1038,3338],[892,3338],[892,3454],[601,3454],[601,3338],[164,3338],[164,3454],[18,3454],[18,3223],[309,3223],[309,2993],[-20,2993],[-20,916],[18,916],[18,1031],[309,1031],[309,916],[455,916],[455,800],[601,800],[601,570],[746,570],[746,338],[892,338],[892,223],[1038,223],[1038,107],[892,107],[892,-20],[4116,-20]],
[[455,2416],[601,2416],[601,2301],[455,2301],[455,2416]],
[[455,1955],[601,1955],[601,1839],[746,1839],[746,1493],[455,1493],[455,1955]],
[[1766,1839],[1911,1839],[1911,1724],[1766,1724],[1766,1839]],
[[1911,1724],[2057,1724],[2057,1609],[1911,1609],[1911,1724]],
[[746,1493],[892,1493],[892,1378],[746,1378],[746,1493]],
[[2057,1378],[2203,1378],[2203,1262],[2057,1262],[2057,1378]],
[[1766,1147],[1620,1147],[1620,1262],[1911,1262],[1911,1031],[1766,1031],[1766,1147]],
[[746,1262],[892,1262],[892,1147],[746,1147],[746,1262]],
[[1911,107],[1620,107],[1620,570],[1475,570],[1475,685],[1620,685],[1620,800],[1475,800],[1475,916],[1766,916],[1766,570],[2203,570],[2203,107],[2057,107],[2057,-8],[1911,-8],[1911,107]],
[[1911,1031],[2203,1031],[2203,916],[1911,916],[1911,1031]],
[[3077,454],[2931,454],[2931,338],[2640,338],[2640,685],[2348,685],[2348,800],[2203,800],[2203,916],[2640,916],[2640,800],[2931,800],[2931,685],[3222,685],[3222,570],[3368,570],[3368,454],[3513,454],[3513,338],[3077,338],[3077,454]],
[[2203,107],[2348,107],[2348,223],[2494,223],[2494,-8],[2203,-8],[2203,107]]
]
```

### PR DESCRIPTION
**Pull request by mourner**
Closes #119. When performing hole elimination, there was a rare edge case where there were multiple candidate points for connection that are all visible to the hole point, but some lead to incorrect winding when connected. This PR should make sure the topology is correct when picking connection points. 

I'm not entirely sure this is the best fix, but it fixes the bug we encountered in real data while not regressing the test suite, so I'm inclined to merge.

cc @mrgreywater

### CODE REVIEW

File: src/earcut.js [hunk 1]
```diff
@@ -344,7 +344,8 @@ function findHoleBridge(hole, outerNode) {
 
             tan = Math.abs(hy - p.y) / (hx - p.x); // tangential
 
-            if ((tan < tanMin || (tan === tanMin && p.x > m.x)) && locallyInside(p, hole)) {
+            if ((tan < tanMin || (tan === tanMin && (p.x > m.x || area(m, m.next, p.next) > 0))) &&
```
Comment (mrgreywater) [inline src/earcut.js]: I think the issue is that if the tangent is vertical (tanMin == Inf), and there are multiple coincident possible bride points, there is an issue, because it chooses the first one, instead of the closest. Instead of `p.x > m.x`, it should probably check `p.x*p.x+p.y*p.y > m.x*m.x+m.y*m.y`. I'll look into it later

Comment (mourner) [reply src/earcut.js]: No, it's not that. See this picture of when it fails:

image

For the last hole, there are multiple nodes in the same spot, with edges going into different directions. We're picking the node that belonged to the second hole (producing invalid connection), rather than picking the node that has the diagonal edge. So, to fix that, for multiple candidates with the same exact coordinates, we want to pick one with an edge closest to the hole point when going counter-clockwise. So I think the fix direction is right, I'm just not confident it covers all similar cases. But better than nothing.

**Comment by mourner** [comment]
Hold on, this fix introduces some regressions — will share more test cases...

**Comment by mourner** [comment]
@mrgreywater here's a set of clean polygons that are failing for `master` version (some of them fail on this branch too, which in turn produces many more failures that are not present on `master`): https://gist.github.com/mourner/59b29d4a62ab2864eaed1a7014f6eb0d

I'm fried for today but will keep cracking on this tomorrow... Any help appreciated though.

**Comment by mourner** [comment]
OK, I've force-pushed the PR to a fix I think is the right one for this particular problem, after fixes in #122 and #121. This brings the number of failed triangulations in my 190k test cases down from 14 to 4.

**Comment by mourner** [comment]
Merging now because we need this fix urgently in production, but I'll be glad to accept any PRs that improve / simplify the fix while keeping the tests happy.

**Comment by mrgreywater** [comment]
I have looked at these PRs and while I can't see anything wrong with them per se, I don't like the fact that I can't really judge the impact on a all edge cases. If you tested them all on a 190k dataset and see an improvement, I'm all for it.

Personally I think we should have some kind of pseudo-random test case generator, so we can get some idea if we're working in the right direction.

**Comment by mourner** [comment]
@mrgreywater yeah, I understand — sorry for the rapid changes! 

Alternatively we could establish a set of real world test cases we can use — e.g. take some very difficult open dataset like terrain contours, slice it into vector tiles with [Tippecanoe](https://github.com/mapbox/tippecanoe) with a defined set of options (so that we multiply the cases with clipping and simplification / grid snapping across zoom levels), and run [earcut-reduce](https://github.com/mapbox/earcut-reduce) against it (the tool I currently use on that private dataset I can't share in full).
